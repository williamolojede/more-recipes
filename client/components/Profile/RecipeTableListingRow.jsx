import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import { recipePropTypes } from '../../config/proptypes';

const propTypes = {
  type: PropTypes.oneOf(['favorites', 'personal-recipes']).isRequired,
  asOwner: PropTypes.bool.isRequired
};

const RecipeTableListingRow = ({
  recipe,
  asOwner,
  type,
  removeRecipe,
  index
}) => (
  <tr>
    <td className="my-recipe__manage--table___name">
      <Link to={`/recipe/${recipe.id}`}>
        {recipe.name}
      </Link>
    </td>
    {
      asOwner &&
        <td className="my-recipe__manage--table___action">
          {
            type === 'favorites' ||
            <Link
              to={{
                pathname: '/user/1/modify-recipe',
                state: { recipe, index }
              }}
            >
              <i className="mdi mdi-wrench" />
            </Link>
          }
          <button onClick={e => removeRecipe(e, recipe.id, type, index)}>
            <i className="mdi mdi-delete" />
          </button>
        </td>
    }
  </tr>
);

RecipeTableListingRow.propTypes = {
  ...propTypes,
  recipe: PropTypes.shape(recipePropTypes).isRequired,
  removeRecipe: PropTypes.func.isRequired,
  index: PropTypes.number.isRequired
};

export default RecipeTableListingRow;
