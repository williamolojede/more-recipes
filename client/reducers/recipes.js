/**
 * Reducer for All Recipe
 */

import { RECIEVE_TOP_RATED_RECIPE } from '../actions/types';

const recipes = (state = [], action) => {
  switch (action.type) {
    case RECIEVE_TOP_RATED_RECIPE:
      return action.recipes;
    default:
      return state;
  }
};


export default recipes;
