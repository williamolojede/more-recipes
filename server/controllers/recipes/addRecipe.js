import { Recipe } from '../../models/index';

const addRecipe = (req, res, next) => {
  const { name, description, img_url, ingredients, instructions } = req.body;
  Recipe.create({
    name,
    description,
    img_url,
    ingredients,
    instructions,
    owner: req.userID
  })
    .then(recipe => res.status(201).json({ recipe, message: 'success' }))
    .catch((error) => {
      const err = new Error(error);
      err.status = 500;
      next(err);
    });
};

export default addRecipe;
