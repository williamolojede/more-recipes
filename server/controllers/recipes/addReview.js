import { Review } from '../../models/index';
import helpers from '../../helpers';
import { findReviews } from './getReviews';

const addReview = (req, res, next) => {
  const { content } = req.body;
  const { userID: userId } = req;
  const { id: recipeId } = req.params;

  Review.create({ userId, recipeId, content })
    .then(() => findReviews(req))
    .then(({ rows: reviews, pagination }) => res.status(201).send({
      status: 'success',
      reviews,
      pagination,
      message: 'Your review has been recorded'
    }))
    .catch(error => helpers.systemErrorHandler({
      msg: error.errors[0].message,
      code: 400
    }, next));
};

export default addReview;

