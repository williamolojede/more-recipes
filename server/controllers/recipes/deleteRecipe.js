import { Recipe } from '../../models/index';

const deleteRecipe = (req, res, next) => {
  const { userID } = req;
  // const decoded = jwt.decode(token);
  Recipe.findById(req.params.id)
    .then((recipe) => {
      if (userID !== recipe.owner) {
        const err = new Error('Not authorized to delete this recipe');
        err.status = 403;
        return next(err);
      }
      recipe.destroy()
        .then(() => res.status(200).send({ message: 'success' }))
        .catch(err => console.error(err));
    })
    .catch((error) => {
      const err = new Error(error);
      err.status = 401;
      next(err);
    });
};

export default deleteRecipe;
