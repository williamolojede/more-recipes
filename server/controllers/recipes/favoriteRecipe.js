import { Favorite } from '../../models/index';

const favoriteRecipe = (req, res, next) => {
  const { userID, recipeOwner } = req;
  const recipeId = req.params.id;

  // ensures owner doesn't favorite their own recipe
  if (recipeOwner === userID) {
    const err = new Error('you are not allowed to vote on your own recipe');
    err.status = 403;
    return next(err);
  }

  Favorite.findAll({ where: { recipeId } })
    .then((favorites) => {
      // check if the favorite table is empty i:e its the first favorite for the recipe
      if (favorites.length === 0) {
        return Favorite.create({
          userId: userID,
          recipeId
        })
          .then(() => res.status(200).send({ status: 'success', message: 'recipe added to your favorite list' }))
          .catch((error) => {
            const err = new Error(error);
            err.status = 500;
            return next(err);
          });
      }
      // get list of all userId that has already favorited the recipe
      const alreadyfavorited = [];
      favorites.forEach((el) => {
        alreadyfavorited.push(el.dataValues.userId);
      });
      const userFavorite = favorites.filter(el => el.dataValues.userId === userID)[0];
      // if current user already favorited => remove
      if (alreadyfavorited.includes(userID)) {
        return Favorite.findById(userFavorite.dataValues.id)
          .then(favorite => favorite.destroy())
          .then(() => res.status(200).send({ message: 'recipe removed from your favorite list', status: 'success' }));
      }
      Favorite.create({
        userId: userID,
        recipeId
      })
        .then(() => res.status(200).send({ status: 'success', message: 'recipe added to your favorite list' }))
        .catch((error) => {
          const err = new Error(error);
          err.status = 500;
          return next(err);
        });
    });
};

export default favoriteRecipe;
