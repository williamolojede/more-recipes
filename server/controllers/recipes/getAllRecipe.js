import { Recipe, Review, User } from '../../models/index';
import systemErrorHandler from '../../helpers/systemErrorHandler';
import Paginate from '../../helpers/paginate';

const getAllRecipe = (req, res, next) => {
  const { sort, order } = req.query;

  //  if query is passed
  if (sort && order) {
    // if the wrong sort order is passed
    if (order !== 'ascending' && order !== 'descending') {
      const err = new Error('invalid sort order');
      err.status = 400;
      return next(err);
    }
    return Recipe.findAll({
      order: [['upVoteCount', 'DESC']],
      include: [
        { model: Review, as: 'reviews' },
        { model: User, attributes: ['id', 'username', 'fullname'] }
      ]
    })
      .then((topRecipes) => {
        const {
          recipes,
          metaData
        } = (new Paginate(topRecipes, parseInt(req.query.limit, 10)))
          .getRecipesForPage(parseInt(req.query.page, 10));
        res.status(200).send({ recipes, metaData, message: 'success' });
      })
      .catch(error => systemErrorHandler(error, next));
  }
  Recipe.findAll({
    include: [
      { model: Review, as: 'reviews' },
      { model: User, attributes: ['id', 'username', 'fullname'] }
    ]
  })
    .then(recipes => res.status(200).send({ recipes, message: 'success' }))
    .catch(error => systemErrorHandler(error, next));
};
export default getAllRecipe;
