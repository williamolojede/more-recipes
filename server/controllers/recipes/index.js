import createRecipe from './addRecipe';
import getAllRecipe from './getAllRecipe';
import updateRecipe from './updateRecipe';
import deleteRecipe from './deleteRecipe';

const recipesController = {
  getAllRecipe,
  createRecipe,
  updateRecipe,
  deleteRecipe
};


export default recipesController;
