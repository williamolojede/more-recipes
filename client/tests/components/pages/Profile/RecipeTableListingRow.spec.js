import React from 'react';
import RecipeTableListingRow from '../../../../components/pages/Profile/RecipeTableListingRow';
import recipes from '../../../../__mocks__/recipes.mock';

function setup(type, asOwner) {
  const props = {
    type,
    asOwner,
    recipe: recipes[0],
    removeRecipe: jest.fn(),
    index: 1
  };
  const wrapper = shallow(<RecipeTableListingRow {...props} />);

  return {
    wrapper,
    props
  };
}


describe('RecipeTableListingRowu', () => {
  it('should render with right amount of elements for favorite recipes', () => {
    const { wrapper } = setup('favorites', true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with right amount of elements for personal-recipe', () => {
    const { wrapper } = setup('personal-recipes', true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should dispatch an action to log user out', () => {
    const { wrapper, props } = setup('personal-recipes', true);
    const deleteButton = wrapper.find('button');
    const dispatchSpy = jest.spyOn(wrapper.instance().props, 'removeRecipe');
    deleteButton.simulate('click', eventObject);
    expect(dispatchSpy).toHaveBeenCalledTimes(1);
    expect(dispatchSpy)
      .toHaveBeenCalledWith(eventObject, props.recipe.id, props.type, props.index);
  });
});
