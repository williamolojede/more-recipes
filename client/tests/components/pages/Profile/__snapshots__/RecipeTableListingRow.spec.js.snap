// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeTableListingRowu should render with right amount of elements for favorite recipes 1`] = `
<tr>
  <td
    className="my-recipe__manage--table___name"
  >
    <Link
      replace={false}
      to="/recipe/2"
    >
      Jollof Rice
    </Link>
  </td>
  <td
    className="my-recipe__manage--table___action"
  >
    <button
      onClick={[Function]}
    >
      <i
        className="mdi mdi-delete"
      />
    </button>
  </td>
</tr>
`;

exports[`RecipeTableListingRowu should render with right amount of elements for personal-recipe 1`] = `
<tr>
  <td
    className="my-recipe__manage--table___name"
  >
    <Link
      replace={false}
      to="/recipe/2"
    >
      Jollof Rice
    </Link>
  </td>
  <td
    className="my-recipe__manage--table___action"
  >
    <Link
      replace={false}
      to={
        Object {
          "pathname": "/user/1/modify-recipe",
          "state": Object {
            "index": 1,
            "recipe": Object {
              "User": Object {
                "fullname": "example user",
                "id": 2,
                "username": null,
              },
              "createdAt": "2017-10-23T08:08:47.566Z",
              "description": "It is simply better than ghanian jollof",
              "downVoteCount": 1,
              "favoriteCount": 3,
              "id": 2,
              "img_url": "http://www.africanbites.com/wp-content/uploads/2014/05/IMG_9677-2-1-150x150.jpg",
              "ingredients": Array [
                "rice",
                "canola oil",
              ],
              "instructions": Array [
                "Break up the clumpy rice before starting.",
                "Garnish with chopped scallion and serve",
              ],
              "name": "Jollof Rice",
              "owner": 2,
              "upVoteCount": 3,
              "updatedAt": "2018-01-16T10:02:13.814Z",
              "viewCount": 6,
            },
          },
        }
      }
    >
      <i
        className="mdi mdi-wrench"
      />
    </Link>
    <button
      onClick={[Function]}
    >
      <i
        className="mdi mdi-delete"
      />
    </button>
  </td>
</tr>
`;
